### PR TITLE
Updating text field location in IOS as a pre-work for spellcheck

### PR DESCRIPTION
**Pull request by nturgut**
The location/transform of the text input is updated at post frame for all the browsers except IOS (with Webkit engine). 

In IOS this update was only on touch end. When spellcheck is enabled the red underling was static and were not moving with scroll.

This PR changes the location updates which is pre-work for enabling spellcheck.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/dom_renderer.dart [hunk 1]
```diff
@@ -316,22 +316,7 @@ flt-glass-pane * {
...
-    /// Flutter web is positioning the input box on top of editable widget.
-    /// This location is updated only in the paint phase of the widget.
-    /// It is wrong during the scroll. It is not important for text editing
-    /// since the content is already invisible. On the other hand, the red
-    /// indicator for spellcheck gets confusing due to the wrong positioning.
-    /// We are disabling spellcheck until the location starts getting updated
-    /// via scroll. This is possible since we can listen to the scroll on
-    /// Flutter.
-    /// See [HybridTextEditing].
-    bodyElement.spellcheck = false;
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/dom_renderer.dart]: We also do this in semantics text field. Do we wanna remove those as well?

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 2]
```diff
@@ -741,6 +800,19 @@ class HybridTextEditing {
...
+  // TODO(flutter_web): When the browser closes and re-opens the virtual
+  // shifts the page in IOS. Call this method from visibility change listener
+  // attached to body.
+  /// Set the dom elements location somewhere outside of the screen.
+  ///
+  /// This is useful for not triggering a scroll when IOS virtual keyboard is
+  /// coming up.
+  ///
+  /// See [TextEditingElement._delayBeforePositining].
+  void setStyleOutsideOfScreen(html.HtmlElement domElement) {
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing.dart]: Is there a reason for this method to be in `HybridTextEditing`?

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 3]
```diff
@@ -715,10 +768,16 @@ class HybridTextEditing {
...
+  /// This method changes the location of the input element if it is already
+  /// positioned.
+  /// See [TextEditingElement._delayBeforePositining].
   void _setDynamicStyleAttributes(html.HtmlElement domElement) {
     if (_editingLocationAndSize != null &&
-        !(browserEngine == BrowserEngine.webkit &&
-            operatingSystem == OperatingSystem.iOs)) {
+        (inputPositioned ||
+            !(browserEngine == BrowserEngine.webkit &&
+                operatingSystem == OperatingSystem.iOs))) {
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing.dart]: This condition is becoming harder to understand. I suggest adding getters to make it more readable:
```dart
bool get _canPositionInput => inputPositioned || _doesKeyboardShiftInput;
bool get _doesKeyboardShiftInput =>
    browserEngine == BrowserEngine.webkit && operatingSystem == OperatingSystem.iOs;

void _setDynamicStyleAttributes(html.HtmlElement domElement) {
  if (_editingLocationAndSize != null && _canPositionInput) {
    setStyle(domElement);
  }
}
```

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 4]
```diff
@@ -583,8 +625,19 @@ class HybridTextEditing {
...
+  bool get inputElementNeedsToBePositioned =>
+      !inputPositioned &&
+      _isEditing &&
+      browserEngine == BrowserEngine.webkit &&
+      operatingSystem == OperatingSystem.iOs;
+
+  /// Flag indication wheterher the input element's position is set.
+  ///
+  /// See [inputElementNeedsToBePositioned].
+  bool inputPositioned = false;
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing.dart]: It feels like many of the fields here really belong to the `TextEditingElement` class (e.g. `inputPositioned`, `inputElementNeedsToBePositioned`, `_editingLocationAndSize`, etc).

Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Thanks for the suggestion. I agree, I'll make a refactor in the next PR.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 2 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Agreed, I'll also carry this method during the refactor.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 5]
```diff
@@ -203,6 +203,22 @@ class TextEditingElement {
...
+  /// has shifted. Therefore instead of listening to the shift and passing this
+  /// information to Flutter Framework, we are trying to stop the shift.
+  ///
+  /// In IOS, the virtual keyboard shifts the screen up if the focused input
+  /// element is under the keyboard or very close to the keyboard. Before the
+  /// focus is called we are positining it offscreen. The location of the input
+  /// in IOS is set to correct place, 100ms after focus. We use this timer for
+  /// timing this delay.
+  Timer _positionInputElementTimer;
+  static const Duration _delayBeforePositining =
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done! Thanks

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 3 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Thank you, it looks a lot better now. 

(note: I defined as _canPositionInput => inputPositioned || !_doesKeyboardShiftInput)

File: lib/web_ui/lib/src/engine/dom_renderer.dart (hunk 1 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/dom_renderer.dart]: I decided  not to enable spellcheck in this cl. I am just making positioning related changes necessary as a preliminary work for spellcheck.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 6]
```diff
@@ -741,6 +807,19 @@ class HybridTextEditing {
       ..transform = transformCss;
   }
 
+  // TODO(flutter_web): When the browser closes and re-opens the virtual
+  // shifts the page in IOS. Call this method from visibility change listener
+  // attached to body.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing.dart]: Why is the above comment a `TODO`?

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/text_editing.dart]: See also: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-checking-in-comments-that-ask-questions

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 7]
```diff
@@ -741,6 +807,19 @@ class HybridTextEditing {
       ..transform = transformCss;
   }
 
+  // TODO(flutter_web): When the browser closes and re-opens the virtual
+  // shifts the page in IOS. Call this method from visibility change listener
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing.dart]: nit: it's spelled iOS (applies here and elsewhere in this PR)

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 8]
```diff
@@ -741,6 +807,19 @@ class HybridTextEditing {
...
+  // attached to body.
+  /// Set the dom elements location somewhere outside of the screen.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing.dart]: nit: element's

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 6 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Cause it will take anoter 50+lines. This PR is already to big. It is a TODO for a next PR :)

Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: It is not a question by the way, when meant as a conjunction. I changed when with after.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 9]
```diff
@@ -222,20 +238,22 @@ class TextEditingElement {
...
   ///
-  /// This method is called in the end of the 'touchend' event, therefore it is
-  /// called after the editing state is set.
+  /// This method is called after a delay.
+  /// See [_positionInputElementTimer].
   void configureInputElementForIOS() {
     if (browserEngine != BrowserEngine.webkit ||
         operatingSystem != OperatingSystem.iOs) {
-      // Only relevant on Safari.
+      // Only relevant on Safari Mobile and Chrome on IOS.
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 10]
```diff
@@ -274,6 +292,31 @@ class TextEditingElement {
...
+        _positionInputElementTimer?.cancel();
+        _positionInputElementTimer = Timer(_delayBeforePositining, () {
+          if (textEditing.inputElementNeedsToBePositioned) {
+            configureInputElementForIOS();
+          }
+        });
+
+        /// When the virtual keyboard is closed on IOS, onBlur is triggered.
+        _subscriptions.add(domElement.onBlur.listen((_) {
+          /// Cancel the timer since there is no need to set the location of the
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 11]
```diff
@@ -274,6 +292,31 @@ class TextEditingElement {
...
+          /// input element anymore. It needs to be focused again to be editable
+          /// by the user.
+          _positionInputElementTimer?.cancel();
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 8 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 7 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 12]
```diff
@@ -274,6 +292,31 @@ class TextEditingElement {
...
+      /// See [_positionInputElementTimer].
+      owner.setStyleOutsideOfScreen(domElement);
+
+      _subscriptions.add(domElement.onFocus.listen((_) {
+        // Cancel previous timer if exists.
+        _positionInputElementTimer?.cancel();
+        _positionInputElementTimer = Timer(_delayBeforePositining, () {
+          if (textEditing.inputElementNeedsToBePositioned) {
+            configureInputElementForIOS();
+          }
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 13]
```diff
@@ -274,6 +292,31 @@ class TextEditingElement {
       }));
     }
 
+    if (owner.doesKeyboardShiftInput) {
+      /// Position the element outside of the page before focusing on it.
+      ///
+      /// See [_positionInputElementTimer].
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 14]
```diff
@@ -710,15 +761,30 @@ class HybridTextEditing {
...
   }
 
+  /// Positioning of input element is only done if we are not expecting input
+  /// to be shifted by a virtual keyboard or if the input is already positioned.
+  ///
+  /// Otherwise positioning will be done after focusing on the input.
+  /// See [TextEditingElement._delayBeforePositining].
+  bool get _canPositionInput => inputPositioned || !doesKeyboardShiftInput;
+
+  /// Flag indicating if virtual keyboard shifts the location of input element.
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 15]
```diff
@@ -741,6 +807,19 @@ class HybridTextEditing {
       ..transform = transformCss;
   }
 
+  // TODO(flutter_web): When the browser closes and re-opens the virtual
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

**Comment by nturgut** [comment]
Thanks for all the comments @mdebbar  @yjbanov

**Comment by nturgut** [comment]
I chencked LUCI, it was failing for unrelated reasons

**Pull request merged**